build/src/lib.rs

```diff
@@ -297,43 +466,50 @@ impl BinaryBuilder {
     }
 
     /// Builds by {downloading, copying, building} the binaries to the out directory.
-    pub fn build(mut self) -> Self {
-        for lang_code in self.language_codes[..].to_vec() {
-            self.build_language(&lang_code);
-        }
-        self
+    pub fn build(mut self) -> Result<Self> {
+        self.language_codes
+            .clone()
+            .into_iter()
+            .try_for_each(|lang_code| self.build_language(&lang_code))?;
+        Ok(self)
     }
 
     /// Validates the binaries by checking if they can be loaded by nlprule.
-    pub fn validate(self) -> Self {
+    pub fn validate(self) -> Result<Self> {
         for lang_code in &self.language_codes {
             let tokenizer_out = self.out_dir.join(tokenizer_filename(lang_code));
             let rules_out = self.out_dir.join(rules_filename(lang_code));
 
-            nlprule::Rules::new(rules_out).unwrap_or_else(|e| {
-                panic!(
-                    "failed to validate rules binary for {lang_code}: {error}",
-                    lang_code = lang_code,
-                    error = e
-                )
-            });
-            nlprule::Tokenizer::new(tokenizer_out).unwrap_or_else(|e| {
-                panic!(
-                    "failed to validate tokenizer binary for {lang_code}: {error}",
-                    lang_code = lang_code,
-                    error = e
-                )
-            });
+            nlprule::Rules::new(rules_out)
+                .map_err(|e| Error::ValidationFailed(lang_code.to_owned(), Binary::Rules, e))?;
+            nlprule::Tokenizer::new(tokenizer_out)
+                .map_err(|e| Error::ValidationFailed(lang_code.to_owned(), Binary::Tokenizer, e))?;
         }
 
-        self
+        Ok(self)
     }
 
     /// Gets the paths to all files this builder created.
     pub fn outputs(&self) -> &[PathBuf] {
         &self.outputs
     }
 
+    /// Applies the given transformation before placing a binencoded file into the cache
+    /// Allows for easier compression usage.
+    /// Modifies the cached path by the given path function.
```

Comment:
Should be a bit more explicit here about when the transformation is called (not only before cache)

```suggestion
    /// Applies the given transformation function to the binary immediately after obtaining it.
    /// This happens before placing the file in the cache (if any) so by using a compression
    /// function the size of the cache directory can be reduced.
    /// Modifies the path of the cached binaries by the given `path_fn`.
    /// If no cache directory is set or the binaries are built from the build dir, the `path_fn` does nothing.
```